File: hashing/check_for_existence/c++/destination_city/destination_city.cpp
// destination_city.cpp
#include "destination_city.hpp"
#include <vector>
#include <string>
#include <unordered_map>

std::string destCity(const std::vector<std::vector<std::string>> &paths)
{
    std::unordered_map<std::string, int> destCounts;
    // O(n) to build a hash table mapping city name -> count
    for (const auto &path : paths)
    {
        const std::string &dep = path[0];
        const std::string &dest = path[1];

        // Insert a new key-value in place if there is no element with the key
        // The 'result.first' is an iterator pointing to the inserted element
        // The 'result.second' is true if the insertion took place or false if key already exists
        auto result = destCounts.emplace(dep, -1);
        // If the insertion didn't happen
        if (!result.second)
        {
            // Decrement the value (destination count) of a city
            --result.first->second;
        }

        result = destCounts.emplace(dest, 1);
        // If the insertion didn't happen,
        if (!result.second)
        {
            // Increment the value (destination count) of a city
            ++result.first->second;
        }
    }

    for (const auto &entry : destCounts)
    {
        // If the value, i.e., destination count, is > 0
        if (entry.second > 0)
        {
            // Return name of city
            return entry.first;
        }
    }
    // Never gets here if a solution is guaranteed
    return " ";
}

File: hashing/check_for_existence/python3/destination_city.py
from collections import defaultdict
from typing import List


def dest_city(paths: List[List[str]]) -> str:
    """
    Given an array of paths, where `paths[i] = [cityA_i, cityB_i]` means there exists a direct path going from `cityA_i` to `cityB_i`. Return the destination city, 
    that is, the city without any path outgoing to another city.
    
    Note: It is guaranteed that the graph of paths forms a line without any loop, therefore, there will be exactly one destination city.

    Parameters
    ----------
    paths : List[List[str]]
        A list of lists of strings, each of which is a dep -> dest pair of cities

    Returns
    -------
    str
        Name of the destination city
    """
    dest_count = defaultdict(int)
    # Cost O(n) to build
    for dep, dest in paths:
        # Decrement if city is a departure
        dest_count[dep] -= 1
        # Increment if city is a destination
        dest_count[dest] += 1
    
    # Find the city that has > 0 count, that is the destination
    for dep, dest in paths:
        # Access is O(1)
        if dest_count[dest] > 0:
            return dest
    # Never gets here if solution is guaranteed
    return ''
    
    
def main() -> int:
    
    paths = [["London", "New York"], ["New York", "Lima"], ["Lima", "Sao Paulo"]]
    destination = dest_city(paths)
    print(f"Given paths {paths}, the destination city is {destination}")
    
    return 0
    
if __name__ == "__main__":
    
    main()
